optuna/storages/in_memory.py

```diff
@@ -53,126 +52,160 @@ def __setstate__(self, state):
     def create_new_study(self, study_name=None):
         # type: (Optional[str]) -> int
 
-        if study_name is not None:
-            self.study_name = study_name
+        with self._lock:
+            study_id = self._max_study_id + 1
+            self._max_study_id += 1
+
+            if study_name is not None:
+                if study_name in self._study_name_to_id:
+                    raise DuplicatedStudyError
+            else:
+                study_uuid = str(uuid.uuid4())
+                study_name = DEFAULT_STUDY_NAME_PREFIX + study_uuid
+            self._studies[study_id] = _StudyInfo(study_name)
+            self._study_name_to_id[study_name] = study_id
 
-        return IN_MEMORY_STORAGE_STUDY_ID  # TODO(akiba)
+            return study_id
 
     def delete_study(self, study_id):
         # type: (int) -> None
 
-        self._check_study_id(study_id)
-
         with self._lock:
-            self.trials = []
-            self.best_trial_id = None
-            self.param_distribution = {}
-            self.direction = StudyDirection.NOT_SET
-            self.study_user_attrs = {}
-            self.study_system_attrs = {}
-            self.study_name = DEFAULT_STUDY_NAME_PREFIX + IN_MEMORY_STORAGE_STUDY_UUID
+            self._check_study_id(study_id)
+
+            with self._lock:
```

Comment:
We should not nest a lock acquisition here. There are a couple of similar occurrences further down.

Reply:
Exactly. I wonder why these did not break the test. (These methods might not be covered in old tests.)